controller/nodelifecycle: Make monitorNodeHealth process nodes concurrently

cbroglie opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Marking the pods not ready on a node requires looping over them and updating each pod's status one at a time. This is performed serially, and can take a while if we're processing each node serially as well.

Since the time is spent waiting on io, there's an opportunity to go faster by processing multiple nodes concurrently. This change modifies the loop to process nodes in parallel, using the same number of workers as `doNodeProcessingPassWorker`.

This change also introduces histogram metrics to better observe `monitorNodeHealth`.

#### Which issue(s) this PR fixes:

ref #114295

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

aojea commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -824,71 +825,98 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
-				if err := nc.processNoTaintBaseEviction(ctx, node, &observedReadyCondition, gracePeriod, pods); err != nil {
-					utilruntime.HandleError(fmt.Errorf("unable to evict all pods from node %v: %v; queuing for retry", node.Name, err))
+				name := node.Name
+				node, err = nc.kubeClient.CoreV1().Nodes().Get(ctx, name, metav1.GetOptions{})
```
this will mean more concurrent request in parallel 👀

aojea commented:
/assign @aojea @wojtek-t @SergeyKanzhelev @bobbypage 
/ok-to-test

we should have a test that evidences the performance improvement, I didn't check in detail, but I can see that we are doing some Get inside the loop and those may cause some problem, another thing we should check if is there can be races due to the parallelization.

Looking at the metrics I can't see any histogram, I wonder if we can add some metrics too, so we can measure these performance problems instead of relying on the logs

cbroglie commented:
/retest

cbroglie commented:
> Looking at the metrics I can't see any histogram, I wonder if we can add some metrics too, so we can measure these performance problems instead of relying on the logs

Good idea. We have a mini framework based on kubevirt that let's us simulate losing a bunch of nodes that we've been using to measure the impact of this change, and having to extract the timings from the logs has been tedious. I'll look into adding a histogram, and can share some hard numbers after that.

On the surface the outstanding test failures do not appear to be related to this change, but I'll dig in a little deeper too.

helayoty commented on vendor/golang.org/x/sync/semaphore/semaphore.go:
```diff
@@ -0,0 +1,136 @@
+// Copyright 2017 The Go Authors. All rights reserved.
```
nit: We prefer to have the vendor/autogenerated files in a separate commit. It should be ok as the new files are not too many.

helayoty commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -824,71 +830,103 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+	// can take a while if we're processing each node serially as well. So we
+	// process them with bounded concurrency instead, since most of the time is
+	// spent waiting on io.
+	sem := semaphore.NewWeighted(scheduler.UpdateWorkerSize)
```
A unit test is needed to check different worker sizes.

soltysh replied:
Why would you pick semaphore over using a separate goroutine which could update pods statuses?

cbroglie replied:
The semaphore package makes it is easy to implement worker pools to work on independent tasks with bounded parallelism, but I'm not tied to it. Could you elaborate on what you had in mind with the separate goroutine approach?

wojtek-t replied:
+1 to separate goroutines.
In general, the "multiple workers per controller" is a common pattern in many controllers. I think that making different controllers as similar as possible has a huge value for developers, as they don't need to learn a separate patterns for each thing they are touching.

Here the pattern is slightly different (i.e. you don't create N workers at the top of level of the controller), instead you create M workers for processing here.
But even for this pattern, we have a utility function. This function is `ParallelizeUntil`:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L46

I would really prefer changing the approach to use this function, instead of introducing semaphore here (which isn't yet used anywhere else in the code).


[At some point we may consider changing the implementation of ParallelizeUntil to use semaphore, but let's not introduce it here.]

cbroglie replied:
Updated to use `ParallelizeUntil`. It dropped in nicely, and also made the diff smaller.

helayoty commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -824,71 +830,103 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
-					nc.nodesToRetry.Store(node.Name, struct{}{})
+		node := nodes[i].DeepCopy()
+		go func() {
+			start := time.Now()
```
We should use the controller [now](https://github.com/kubernetes/kubernetes/blob/00aae4c10c204d258e62a96f30970e9b91b29149/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L269) func instead of `time.Now()` to prevent skew across the cluster.

wojtek-t replied:
Please fix

helayoty commented on pkg/controller/nodelifecycle/metrics.go:
```diff
@@ -79,6 +79,25 @@ var (
+	updateNodeHealthDuration = metrics.NewHistogram(
+		&metrics.HistogramOpts{
+			Subsystem:      nodeControllerSubsystem,
+			Name:           "update_node_health_duration_seconds",
```
Metrics names should be defined as const similar to other ones. i.e. `evictionsTotalKey`

helayoty commented on pkg/controller/nodelifecycle/metrics.go:
```diff
@@ -79,6 +79,25 @@ var (
+	monitorNodeHealthDuration = metrics.NewHistogram(
+		&metrics.HistogramOpts{
+			Subsystem:      nodeControllerSubsystem,
+			Name:           "update_nodes_health_duration_seconds",
```
Same. Metrics names should be defined as const similar to other ones. i.e. evictionsTotalKey

soltysh replied:
I'll defer to @kubernetes/sig-instrumentation-approvers for verdict, but I think the latter metrics should be sufficient on this.

dgrisonnet replied:
I would definitely recommend to at least changing the name of the metrics since they will be quite confusing and prone to errors to end-users.

@soltysh do you mean that this metric should be sufficient or the other one?

My personal opinion is that `update_node_health_duration_seconds` is the most useful one here since it gives us an idea of the usual time it takes for a node health to be updated. As for the metric measuring the time it takes to update the health of all nodes, I don't really like it since it is not very actionable and could easily be inferred from the node-level one.

wojtek-t replied:
>  Whereas the metric measuring the time it takes to update the health of all nodes is not very actionable and could easily be inferred from the node-level one.

It is potentially actionable e.g. that we need more workers to parallelize more.
But I don't feel super strongly about this.

dgrisonnet replied:
I didn't think about that scenario, I guess that make sense. I would just recommend updating the names then so that metrics are easier to distinguish by the users.

cbroglie replied:
> It is potentially actionable e.g. that we need more workers to parallelize more.

This was my thinking, that it would be easiest to understand the overall performance of  `monitorNodeHealth` with a metric that included everything.

> I would just recommend updating the names then so that metrics are easier to distinguish by the users.

I struggled to name these initially, hopefully the updated names are more obvious to end users. Definitely open to suggestions.

helayoty commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -794,6 +795,11 @@ func (nc *Controller) doEvictionPass(ctx context.Context) {
 // This function will taint nodes who are not ready or not reachable for a long period of time.
 func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+	start := time.Now()
+	defer func() {
```
Why do we need to defer the metrics?

cbroglie replied:
It's not strictly required, but it avoids needing to explicitly handle the different early return cases

helayoty commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -824,71 +830,103 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+
+	// Acquire all tokens to block until every goroutine has finished.
+	if err := sem.Acquire(ctx, scheduler.UpdateWorkerSize); err != nil {
+		return fmt.Errorf("failed to acquire semaphore: %w", err)
```
nit: Error message starts with Caps to be consistent with the code base.

soltysh replied:
> nit: Error message starts with Caps to be consistent with the code base.

That's actually not true, all errors can be and most often will be quoted higher up in the chain of invocation, so we prefer to have them lowercased :wink:

helayoty replied:
> > nit: Error message starts with Caps to be consistent with the code base.
> 
> That's actually not true, all errors can be and most often will be quoted higher up in the chain of invocation, so we prefer to have them lowercased 😉

I prefer all messages to start in lowercase to be honest but I also prefer consistency across the controller. But as I mentioned this is just nit, not a blocker at all.

helayoty commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -824,71 +830,103 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+					fallthrough
+				case needsRetry && observedReadyCondition.Status != v1.ConditionTrue:
+					if err = controllerutil.MarkPodsNotReady(ctx, nc.kubeClient, nc.recorder, pods, node.Name); err != nil {
+						utilruntime.HandleError(fmt.Errorf("unable to mark all pods NotReady on node %v: %v; queuing for retry", node.Name, err))
```
nit: Error message starts with Caps to be consistent with the code base.

cbroglie replied:
This line isn't actually part of my change, it's just the whitespace changing

cbroglie commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -794,6 +795,11 @@ func (nc *Controller) doEvictionPass(ctx context.Context) {
 // if not, post "NodeReady==ConditionUnknown".
 // This function will taint nodes who are not ready or not reachable for a long period of time.
 func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+	start := time.Now()
```
There wouldn't be any skew concerns in this use case, but I'll change it to be consistent

soltysh replied:
The main reason for using controller's now function is mostly testing, so that's good thing to follow, as @helayoty mentioned.

wojtek-t replied:
Not fixed yet - please update.

wojtek-t commented on pkg/controller/nodelifecycle/metrics.go:
```diff
@@ -79,6 +79,25 @@ var (
+			Subsystem:      nodeControllerSubsystem,
+			Name:           "update_nodes_health_duration_seconds",
+			Help:           "Duration in seconds for NodeController to update the health of all nodes.",
+			Buckets:        metrics.ExponentialBuckets(0.01, 2, 16), // 10ms -> ~5m
```
I don't think we need 16 buckets. I'm also not convinced we need that big granularity - probably having metrics.ExponentialBuckets(0.001, 6) is good enough...

wojtek-t commented on pkg/controller/nodelifecycle/metrics.go:
```diff
@@ -79,6 +79,25 @@ var (
+			Subsystem:      nodeControllerSubsystem,
+			Name:           "update_node_health_duration_seconds",
+			Help:           "Duration in seconds for NodeController to update the health of a single node.",
+			Buckets:        metrics.ExponentialBuckets(0.001, 2, 16), // 1ms -> ~30s
```
I'm not sure we need that large granularity. Having metrics.ExponentialBuckets(0.001, 4, 10) is probably good enough.

wojtek-t reviewed (commented):
@cbroglie - thanks for opening this PR. Sorry for huge delay (although you already got some comments here).
I'm highly interested in this PR and helping to land it - I added few more comments. PTAL when you will have a chance.

wojtek-t commented:
> we should have a test that evidences the performance improvement, I didn't check in detail, but I can see that we are doing some Get inside the loop and those may cause some problem,

@aojea 
This is a continuation of https://github.com/kubernetes/kubernetes/issues/110341 problems (that one fixed the most important thing, and we have more).
I agree that we need to be careful about races (I will make a deeper pass once the code gets updated), but I would say that the impact and idea behind parallelizing it is clearly needed and helpful.

cbroglie committed:
controller/nodelifecycle: Make monitorNodeHealth process nodes concurrently

Marking the pods not ready on a node requires looping over them and
updating each pod's status one at a time. This is performed serially,
and can take a while if we're processing each node serially as well.

Since the time is spent waiting on io, there's an opportunity to go
faster by processing multiple nodes concurrently. This change modifies
the loop to process nodes in parallel, using the same number of workers
as doNodeProcessingPassWorker.

This change also introduces histogram metrics to better observe
monitorNodeHealth.

cbroglie commented:
@wojtek-t Thanks! I've taken a pass at addressing all the comments and extending the test coverage, please let me know what you think of the latest approach.

wojtek-t commented on pkg/controller/nodelifecycle/metrics.go:
```diff
@@ -30,6 +30,9 @@ const (
 	evictionsNumberKey      = "evictions_number"
 	evictionsTotalKey       = "evictions_total"
+
+	updateNodeHealthKey     = "update_node_health_duration_seconds"
```
Those names looks reasonable to me now - @dgrisonnet - are you ok with the new ones?

dgrisonnet replied:
Yes, it is easier to notice the difference now :+1:

wojtek-t commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -332,6 +332,10 @@ type Controller struct {
 	nodeMonitorGracePeriod time.Duration
 
+	// Number of workers Controller uses to process node monitor health updates.
+	// Defaults to scheduler.UpdateWorkerSize, but can be overridden in tests.
```
I suggest removing the "but can be overridden in tests".

While I don't think we immediately need to expose it now, I can imagine exposing it as a param of NewNodeLifecycleController in the future.

wojtek-t commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -883,12 +902,24 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+	workers := nc.nodeUpdateWorkerSize
+	if workers == 0 {
+		workers = scheduler.UpdateWorkerSize
+	}
```
Let's remove it from here and move the initialization of nodeUpdateWorkerSize field to `NewNodeLifecycleController` function.

wojtek-t commented on pkg/controller/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -824,13 +833,21 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+	var zoneToNodeConditionsLock sync.Mutex
 	zoneToNodeConditions := map[string][]*v1.NodeCondition{}
-	for i := range nodes {
+	updateNodeFunc := func(piece int) {
```
note to myself - do one more time deeper look into locking and races..

wojtek-t replied:
OK - I went through it and it looks fine.

cbroglie commented:
/retest

wojtek-t reviewed (commented):
This looks good to me know - thanks for making this happen.

wojtek-t commented:
/lgtm
/approve

aojea commented:
@wojtek-t do we want to backport this? or is too risky?

wojtek-t commented:
@aojea - given that I didn't hear any huge concerns (it's not a bug, it's rather an optimization), I don't think it deserves to be backported.

aojea commented:
👍 makes sense
